histomicstk/features/ComputeFSDs.py

```diff
@@ -0,0 +1,253 @@
+import collections
+import numpy as np
+import pandas as pd
+from skimage.measure import regionprops
+from skimage.segmentation import find_boundaries
+
+
+def ComputeFSDs(im_label, K=128, Fs=6, Delta=8, rprops=None):
+    """
+    Calculates `Fourier shape descriptors` for each objects.
+
+    Parameters
+    ----------
+    im_label : array_like
+        A labeled mask image wherein intensity of a pixel is the ID of the
+        object it belongs to. Non-zero values are considered to be foreground
+        objects.
+    K : int, optional
+        Number of points for boundary resampling to calculate fourier
+        descriptors. Default value = 128.
+    Fs : int, optional
+        Number of frequency bins for calculating FSDs. Default value = 6.
+    Delta : int, optional
+        Used to dilate nuclei and define cytoplasm region. Default value = 8.
+
+    rprops : output of skimage.measure.regionprops, optional
+        rprops = skimage.measure.regionprops( im_label ). If rprops is not
+        passed then it will be computed inside which will increase the
+        computation time.
+
+    Returns
+    -------
+    fdata: Pandas data frame containing the FSD features for each
+           object/label.
+
+    References
+    ----------
+    .. [1] D. Zhang et al. "A comparative study on shape retrieval using
+       Fourier descriptors with different shape signatures," In Proc.
+       ICIMADE01, 2001.
+    """
+    # List of feature names
+    feature_list = []
+    for i in range(0, Fs):
+        feature_list = np.append(feature_list, 'FSD' + str(i+1))
+
+    # get Label size x
+    sizex = im_label.shape[0]
+    sizey = im_label.shape[1]
+
+    # get the number of objects in Label
+    if rprops is None:
+        rprops = regionprops(im_label)
+
+    # create pandas data frame containing the features for each object
+    numFeatures = len(feature_list)
+    numLabels = len(rprops)
+    fdata = pd.DataFrame(np.zeros((numLabels, numFeatures)),
+                         columns=feature_list)
+
+    # fourier descriptors, spaced evenly over the interval 1:K/2
+    Interval = np.round(
+        np.power(
+            2, np.linspace(0, np.log2(K)-1, Fs+1, endpoint=True)
+        )
+    ).astype(np.uint8)
+
+    for i in range(numLabels):
+        # get bounds of dilated nucleus
+        min_row, max_row, min_col, max_col = \
+            _GetBounds(rprops[i].bbox, Delta, sizex, sizey)
+        # grab label mask
+        lmask = (
+            im_label[min_row:max_row, min_col:max_col] == rprops[i].label
+        ).astype(np.bool)
+        # find boundaries
+        Bounds = np.argwhere(
+            find_boundaries(lmask, mode="inner").astype(np.uint8) == 1
```

Comment:
@slee172 I think you can speedup by calling call `find_boundaries` just once above the loop for the whole label mask. And then get the boundaries of each object inside the loop by an `and` operation of the patches (of size equal to the object bounding box) of label mask and the binary mask returned by of find_boundaries.